src/babashka/curl.clj

```diff
@@ -174,23 +175,39 @@
                   :headers headers
                   :body body
                   :err err
-                  :process (:proc opts)}]
+                  :process process
+                  :exit exit}]
     response))
 
 ;;;; End Response Parsing
 
+(def unexceptional-status?
+  #{200 201 202 203 204 205 206 207 300 301 302 303 304 307})
+
+(defn- should-throw? [response opts]
+  (let [exceptional-status? (not (unexceptional-status? (:status response)))
+        nonzero-exit? (not (zero? (:exit response)))]
+    (and (:throw-exceptions opts)
+         (or exceptional-status? nonzero-exit?))))
+
 (defn request [opts]
   (let [header-file (File/createTempFile "babashka.curl" ".headers")
         opts (assoc opts :header-file header-file)
+        default-opts {:throw-exceptions true}
+        opts (merge default-opts opts)
         args (curl-command opts)
         response (let [response (-> (exec-curl args opts)
                                     (curl-response->map))]
                    (.delete header-file)
-                   response)]
-    (if (:debug opts)
-      (assoc response
-             :command args
-             :options opts)
+                   response)
+        response (if (:debug opts)
+                   (assoc response
+                     :command args
+                     :options opts)
+                   response)
+        stream? (identical? :stream (:as opts))]
+    (if (and (not stream?) (should-throw? response opts))
+      (throw (ex-info "error" response))
```

Comment:
Do you have a preference on what message should get in the `ExceptionInfo`? I put "error" as a filler.

Reply:
Maybe take a look at what clj-http does in case of an exceptional status code? If we throw because of a non-zero exit code, we could just put the curl error message there.